File: RentalModule.Application/QueryHandlers/Tag/GetTags/GetTagsQueryHandler.cs
using RentalModule.Application.Contract.Tags.GetTags;
using RentalModule.Domain.RepositoryInterfaces.CarTag;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.QueryHandlers;

namespace RentalModule.Application.QueryHandlers.Tag.GetTags;

public class GetTagsQueryHandler : IQueryHandler<GetTagsQuery, HandlerResult<GetTagsResponse, IErrorResult>>
{
    private readonly ICarTagQueryRepository _repository;
        
    public GetTagsQueryHandler(ICarTagQueryRepository repository)
    {
        _repository = repository;
    }
        
    public async Task<HandlerResult<GetTagsResponse, IErrorResult>> Handle(GetTagsQuery request, CancellationToken cancellationToken)
    {
        var tags = await _repository.GetCollectionAsync(
            request.OrderBy, 
            request.OrderDirection,
            request.CarOfferId,
            cancellationToken);

        var response = new GetTagsResponse() {
            OrderBy = request.OrderBy,
            OrderDirection = request.OrderDirection,
            Items = tags
        };
            
        return response;
    }
}

File: RentalApi/Application/query_handlers/car_tag/get_tags/get_tags_query_handler.go
package queries

import (
	"context"
	contract "rental-api/Application.contract/tags/get_tags"
	models "rental-api/Domain/models/domestic"
	repository_interfaces "rental-api/Domain/repository_interfaces/car_tag_repository"
	"rental-api/Domain/responses"
)

type GetTagsQueryHandler struct {
	carTagQueryRepository repository_interfaces.CarTagQueryRepository
}

func NewGetTagsQueryHandler(
	carTagQueryRepository repository_interfaces.CarTagQueryRepository,
) *GetTagsQueryHandler {
	return &GetTagsQueryHandler{
		carTagQueryRepository: carTagQueryRepository,
	}
}

func (h *GetTagsQueryHandler) Handle(ctx context.Context, query *GetTagsQuery) (*contract.GetTagsResponse, error) {
	tags, err := h.carTagQueryRepository.GetTagsByCarOfferId(
		query.CarOfferId,
		&query.Sortable,
	)

	if err != nil {
		if err.Error() == "sql: no rows in result set" {
			response := responses.NewResponse[contract.GetTagsResponse](200, "No tags found")
			return &response, nil
		}
		response := responses.NewResponse[contract.GetTagsResponse](500, "Failed to retrieve tags")
		return &response, nil
	}

	if tags == nil {
		tags = &[]models.CarOfferTagModel{}
	}

	return &contract.GetTagsResponse{
		BaseResponse: responses.NewBaseResponse(200, "Tags retrieved successfully"),
		Items:        *tags,
	}, nil
}
